### PR TITLE
WIP (PT story 132571845) ensure that we have good default tags on all pages

### PR DESCRIPTION
**Pull request by nbartowski**
I wrote an initial acceptance test in seo.features which seems to properly fail. Initially I thought implementing the feature would only require adding a before_filter to the application_controller (:set tags per scrum discussion at https://www.youtube.com/watch?v=SoILIjeb02E&feature=youtu.be [pt of interest beginning at 13:30 mark]). However, I discovered that the work already done by @lollypop27 has a more intricate structure which will require changing some of the work she has done. I decided to submit a WIP pull request to keep those interested apprised of the work. I plan to follow up with @lollypop27 on the slack channel to discuss the changes I plan to make before implementing them

### CODE REVIEW

**Comment by nbartowski** [comment]
I've submitted a commit to implement the, "good default tags feature" (PT story: https://www.pivotaltracker.com/n/projects/742821/stories/132571845). Note that currently there is also a title tag inserted in line 8 of the application.html.erb layout file. I believe this line of code is there to supply a title for error pages. I was going to try and refactor this line  of code but my initial solution broke a few more tests than I was expecting so I thought it might best be left for a different feature. Please advise if this pull request should also include refactoring of line 8 (currently each page will have two places which define the <title> tag - which seems to result in the tags being concatenated from an overall title perspective)

Also, the overall structure used to implement default and specific tags has been changed from my initial suggestion. I found that my initial proposal did not work because of the order in which before_filters were called - in short key instance variables were not assigned before being accessed in certain filters.

The current structure is: (1) default tags are assigned in the ApplicationController, (2) other controllers can override the default tags by calling a local before_filter which reassigns the desired tags (currently used in OrganisationsController and VolunteerOpsController), (3) if more granularity is desired, specific controller actions can reassign tags as desired from within the method block.

**Comment by nbartowski** [comment]
This submission resolves the issue concerning titles for error pages mentioned in the last commit (i.e. what was line 8 of the layout application.html.erb). The initial solution that I tried to implement, which resulted in several tests failing, was in fact the correct one in concept. I'm not sure what I did wrong the first time (probably some syntax issue - I'm a fairly novice programmer), but I believe this last commit properly processes forming a page title.

**Comment by tansaku** [comment]
great start @nbartowski - is this still WIP? sorry I haven't had time to give feedback yet

**Comment by nbartowski** [comment]
@tansaku I was thinking that this is sufficient to implement the feature (i.e. no longer WIP status). I believe the functionality will work with the changes in this branch, but you may want me to include extra testing. As I considered what testing should be implemented, my initial thoughts were that there was not much to be gained beyond the cucumber test I included (at least in the default meta tag case). I thought more comprehensive tests might make sense for when the default tags aren't used (i.e. features that implement more specific page titles and descriptions). Let me know your thoughts and if you would like me to add additional testing. Thanks for the help and feedback!

**Comment by tansaku** [comment]
hi @nbartowski - I think the level of testing is fine - my only concern is that setting the meta_tags gets called twice.  Why not just have the controllers override the relevant methods - no need to call meta_tags twice I think.

**Comment by nbartowski** [comment]
@tansaku I initially tried having the controllers override the relevant methods but it didn't work because pertinent instance variables were not yet assigned. For example, if the OrganisationsController :show action is called we ultimately want the meta_tag :title to be the value of @organisation.name. But the @organization variable is assigned in the before_filter method ":set_organization" of the OrganisationsController which is called **after** the before_filter ":set_tags" of the ApplicationController. So when ApplicationController calls the :set_tags method, @organisation.name has not yet been assigned by the OrganisationsController so the meta_tag :title cannot be set by an overriding method from within the Organizations controller because the value of organization.name will not be available. (I hope that explanation was somewhat clear - let me know if not).

The only way I could see to get a method override within OrganisationsController to work was if I assigned the @organisation instance variable within that method. But then that variable will get assigned twice, mimicking the undesired behavior of setting a value twice. I tried to think of some structure that I could use to avoid assigning a value twice, but I think in order to do so would require extensive reorganization of when and where certain instance variables are assigned. Such an endeavor seemed outside the scope of this particular feature.

I'm very interested to hear your feedback, whether there is a solution that I didn't see, or if I should have communicated the inefficiency so that the team could consider making a new feature story to refactor the code for a more efficient solution.

As always, thanks a ton for the feedback and time!

**Comment by nbartowski** [comment]
@tansaku I just realized that an after_filter instead of a before_filter probably would have solved my issue. I'll make the changes and push the implementation, but I probably won't be able to get to it until late monday.

**Comment by nbartowski** [comment]
@tansaku I tried implementing the feature with an after_filter in the ApplicationController but found that it doesn't run in time for the render. This stackoverflow post seems to support my findings (http://stackoverflow.com/questions/9281224/filter-to-execute-before-render-but-after-controller). Please advise how you would like me to proceed. Thanks!

**Comment by lollypop27** [comment]
@nbartowski looks good, thank you.
@tansaku I am happy

**Comment by tansaku** [comment]
@nbartowski I think there are some alternates - can you give me push access on your repo and then I can show you on this branch - or I'll create another ...

**Comment by nbartowski** [comment]
@tansaku I think I've granted you access to my repo. If not, please let me know. Thanks!

**Comment by tansaku** [comment]
@nbartowski - have made simpler with prepend_before_action - take a look and see what you think.

BTW, why did that cucumber step get updated?

from:

```rb
expect(page).to have_description description
```

to

```rb
  expect(page.find(%Q{meta[name="#{name_attribute.downcase}"]}, :visible => false).class).to eq(Capybara::Node::Element)
```

was the former insufficient?

**Comment by tansaku** [comment]
by the way we reviewed the code live at the end of this scrum if you are interested in seeing the process: http://youtu.be/PjpUk3fBBDs :-)

**Comment by nbartowski** [comment]
1) @tansaku thanks for the link to the code review! Yes, I like the changes to ApplicationController, OrganisationController and VolunteerController. Your solution was exactly what I tried but couldn't get to work as I didn't know about the prepend_before_action.

2) However, the change you made to the layout application.html.erb file will break the testing for the static error pages (as Lara mentioned). Those pages set their title in a different manner (using "content_for" method. I did change how the title was set on those pages however in my last commit (https://github.com/AgileVentures/LocalSupport/pull/399/commits/94fc655a4cf3a7dbec83719e6d5df2a28f4076a6). Previously, a second title tag was being inserted into **all pages, not just the static error pages,** by line 8 of the code in application.html.erb. I deleted line 8 and inserted line 6, as the commit shows. The change I made overrides the title tag in application.html.erb if there is content_for? :title. Thus, there will only be one title tag on all pages with the correct title being inserted for the static error pages. I discussed this briefly in the above conversations. I believe this change is why your commit didn't pass the checks.

3) Regarding the change to the cucumber test, the previous:

 expect(page).to have_description description

doesn't work as there is no page.description method. The new cucumber step I wrote merely looks for the presence of a description meta tag, as I figured that was sufficient to ensure the description meta-tag is being set. 

Really appreciate the review and feedback. Very helpful!

**Comment by nbartowski** [comment]
@tansaku I made the edits we discussed, however, when I ran the rspec and cucumber tests I had several new failures (2 rspec failures and 1 cucumber fail; you can see the details of the rspec failure in the "Details" of the "continuous-integration/..." test above. The cucumber test that fails is the second scenario of "custom_error_pages.feature") I investigated some but have not been able to discern what is wrong. Nothing I changed should have affected any of the failing tests. I think it may be related to the code that I pulled in. Note, I first pulled in the develop branch from localsupport and then pulled in your commits to the feature branch of my fork of the repository. At least that is what I think I did. The commit history shows these two items.

Even though I had these issues I wanted to push my latest changes as I won't be able to look at this again until Monday and I thought you might be expecting something sooner. I'll plan on attending the scrum next week if I don't see something sooner.

Thanks!

File: app/controllers/organisations_controller.rb [hunk 1]
```diff
@@ -30,17 +30,21 @@ def index
   def show
     if @organisation.nil?
       render template: 'pages/404', status: 404
+      return
```
Comment (nbartowski) [inline app/controllers/organisations_controller.rb]: the changes in the show method accomplished two purposes: 1) reduce the nesting of "if/else" clauses, and 2) reduces the number of comparisons being made (many lines ended with "if current_user"; I substituted with one comparison and a block)

File: app/controllers/organisations_controller.rb [hunk 2]
```diff
@@ -129,7 +133,9 @@ def self.build params
   private
 
   def set_organisation
-    @organisation = Organisation.friendly.find(params[:id]) rescue nil
+    @organisation = Organisation.friendly.find(params[:id])
+    rescue ActiveRecord::RecordNotFound
+      @organisation = nil
```
Comment (nbartowski) [inline app/controllers/organisations_controller.rb]: the changes in the "set_organisation" method where an attempt to eliminate "codeclimate" issues stating, "avoid using rescue in its modifier form"

File: app/controllers/volunteer_ops_controller.rb [hunk 3]
```diff
@@ -18,6 +18,10 @@ def index
   end
 
   def show
+    if @volunteer_op.nil?
+      render template: 'pages/404', status: 404
+      return
+    end
```
Comment (nbartowski) [inline app/controllers/volunteer_ops_controller.rb]: changes in show method are implemented to satisfy new test in custom_error_pages.feature (mimics functionality added to OrganisationsController in previous commit)

File: app/controllers/volunteer_ops_controller.rb [hunk 4]
```diff
@@ -97,11 +101,15 @@ def current_user_has_organisation?
   end
 
   def set_organisation
-    @organisation = Organisation.friendly.find(params[:organisation_id]) rescue nil
+    @organisation = Organisation.friendly.find(params[:organisation_id])
+    rescue ActiveRecord::RecordNotFound
+      @organisation = nil
```
Comment (nbartowski) [inline app/controllers/volunteer_ops_controller.rb]: I refactored set_organisation method to satisfy "codeclimate" which had issue, "avoid using rescue in its modifier form". Note that rescuing from nil or ActiveRecord::RecordNotFound is probably not necessary as this controller did not make the "set_organisation" a "prepend_before_action" filter as OrganisationsController did. Thus the rescue code is not executed by any test, causing "coverage/coveralls" check to fail.

File: app/controllers/volunteer_ops_controller.rb [hunk 5]
```diff
@@ -97,11 +101,15 @@ def current_user_has_organisation?
...
   end
 
   def set_volunteer_op
     @volunteer_op = VolunteerOp.find(params[:id])
+    rescue ActiveRecord::RecordNotFound
+      @volunteer_op = nil
```
Comment (nbartowski) [inline app/controllers/volunteer_ops_controller.rb]: modified "set_volunteer_op" method to provide needed functionality to pass new cucumber test in "custom_error_pages.feature"

File: features/custom_error_pages.feature [hunk 6]
```diff
@@ -17,6 +17,12 @@ Feature: I want to make error pages follow the general design of the site.
     And the response status should be "404"
     And I should see "We're sorry, but we couldn't find the page you requested"
 
+   Scenario: 404 page when opening an invalid opportunity
+    When I visit "/volunteer_ops/foo-bar-op"
+    And the page should be titled "404 Page Not Found"
+    And the response status should be "404"
+    And I should see "We're sorry, but we couldn't find the page you requested"
+
```
Comment (nbartowski) [inline features/custom_error_pages.feature]: Added scenario to test invalid opportunity, similar to previous scenario which test for 404 page when invalid project (i.e. organisation)

**Comment by nbartowski** [comment]
@tansaku I didn't realize my comments on the commit code would show up in this conversation thread, but I left anyways so you could see the reasoning behind the changes. Let me know if you want to see additional changes to address codeclimate and coverage/coveralls test. Most of the issues could quickly be resolved with the exception of the long method and long class style guide issues.

**Comment by tansaku** [comment]
thanks for all your work on this @nbartowski - I wouldn't worry about the codeclimate issues until we've gotten to the bottom of the error - I hope to look at it in more detail soon ...

**Comment by tansaku** [comment]
@OlenaCh thanks for the review - very much appreciated.

the thing that's currently blocking merging this PR in is this one failing test:

```
  Scenario: Meta Title and Description for volunteers edit page                       # features/admin/seo.feature:44
    Given I am signed in as a superadmin                                              # features/step_definitions/authentication_steps.rb:17
    Then I visit the edit page for the volunteer_op titled "Helping Volunteer"        # features/step_definitions/navigation_steps.rb:89
    Then I should have a page with a title: "Helping Volunteer | Harrow volunteering" # features/step_definitions/basic_steps.rb:574
      expected "" to include "Helping Volunteer | Harrow volunteering" (RSpec::Expectations::ExpectationNotMetError)
      ./features/step_definitions/basic_steps.rb:575:in `/^I should have a page with a title: "([^"]*)"$/'
      features/admin/seo.feature:47:in `Then I should have a page with a title: "Helping Volunteer | Harrow volunteering"'
    And it should have a Meta Description                                             # features/step_definitions/basic_steps.rb:578

Failing Scenarios:
cucumber -p second_try features/admin/seo.feature:44 # Scenario: Meta Title and Description for volunteers edit page
```

**Comment by tansaku** [comment]
@OlenaCh @nbartowski  I think I might have just fixed it with:

```
class VolunteerOpsController < ApplicationController
  before_action :set_tags, only: [:show, :edit]
```

**Comment by OlenaCh** [comment]
@tansaku @nbartowski I am sorry that I am answering only now - I've answered via email yesterday, but it was not published and I did not notice. Actually, I'm pushing my changes to my PR now (if there errors, I'll fix them tomorrow) and then I can help to hunt this bug :).

**Comment by tansaku** [comment]
ugh, feel like I've seen this error before:

```
  Scenario: 404 page when opening an invalid opportunity                        # features/custom_error_pages.feature:20
    When I visit "/volunteer_ops/foo-bar-op"                                    # features/step_definitions/navigation_steps.rb:21
    And the page should be titled "404 Page Not Found"                          # features/step_definitions/navigation_steps.rb:104
      expected "500 Internal Server Error | Harrow volunteering" to include "404 Page Not Found" (RSpec::Expectations::ExpectationNotMetError)
      ./features/step_definitions/navigation_steps.rb:105:in `/^the page should be titled "(.*?)"$/'
      features/custom_error_pages.feature:22:in `And the page should be titled "404 Page Not Found"'
    And the response status should be "404"                                     # features/step_definitions/custom_errors_steps.rb:1
    And I should see "We're sorry, but we couldn't find the page you requested" # features/step_definitions/basic_steps.rb:336
```

thought I'd fixed that already ...

**Comment by nbartowski** [comment]
@tansaku and @OlenaCh Sorry I've been out of the loop for awhile. I glanced briefly at the recent history and it seems that a new error is popping up. Note that my last commit was error free at the time (it only had some style issues with code climate) so a recent PR must have introduced the present failure.

I have a pretty full slate right now, but I will try and grab some time this afternoon to look at the current error.

**Comment by nbartowski** [comment]
@tansaku  and @OlenaCh I was parsing through the commit changes made and realized the error: the logical "and" was switched to "&&" due to a code climate comment. I had done the same in commit "adac7d3":

(https://github.com/AgileVentures/LocalSupport/pull/399/commits/adac7d3694429488884c23b45d63509ee6bc0f8e)

 above and had to switch back in commit "aa16782:

(https://github.com/AgileVentures/LocalSupport/pull/399/commits/aa16782f881c8c6224938584daac1713e1b5a967)

Basically "and" and "&&" are not the same. There is an order of operations difference which causes the error. I'll update my code with the latest develop branch, make the change back, and resubmit

**Comment by nbartowski** [comment]
@tansaku So I believe this last commit resolves the error issues; just codeclimate stuff which I think you were planning to bypass.

**Comment by nbartowski** [comment]
@tansaku my last commit fixes the error; just codeclimate issues left which I think you were going to bypass.

**Comment by tansaku** [comment]
@nbartowski - yeah, good to go now I think - there's some other refactorings but I think we need to get this out ...

**Comment by tansaku** [comment]
@OlenaCh some great ideas in your review - I wonder if we can turn them into refactoring tickets?

**Comment by OlenaCh** [comment]
@tansaku, if you think they are good idea, it can be possible to create two refactoring tickets: one for cucumber and one for controller (or even controllers - look through existing controllers for this certain issue - a lot of variables are created and assigned in actions - and replace it).

**Comment by tansaku** [comment]
@OlenaCh that sounds good - do you have permissions in the tracker? I think you do - please make those tickets and you can start votes on them in the localsupport slack channel so we can get more input from the team on them ...

**Comment by OlenaCh** [comment]
@tansaku yes, I have permissions and I've added two new tickets :). May I ask you how I can start voting, please?
